app/src/main/java/org/astraea/partitioner/smooth/SmoothWeightRoundRobin.java

```diff
@@ -0,0 +1,98 @@
+package org.astraea.partitioner.smooth;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import org.astraea.cost.Periodic;
+import org.astraea.cost.broker.CostUtils;
+
+/**
+ * Given initial key-score pair, it will output a preferred key with the highest current weight. The
+ * current weight of the chosen key will decrease the sum of effective weight. And all current
+ * weight will increment by its effective weight. It may result in "higher score with higher chosen
+ * rate". For example:
+ *
+ * <p>||========================================||============================================||
+ * ||------------ Broker in cluster ------------||------------- Effective weight -------------||
+ * ||------------------Broker1------------------||-------------------- 5 ---------------------||
+ * ||------------------Broker2------------------||-------------------- 1 ---------------------||
+ * ||------------------Broker3------------------||-------------------- 1 ---------------------||
+ * ||===========================================||============================================||
+ *
+ * <p>||===================||=======================||===============||======================||
+ * ||--- Request Number ---|| Before current weight || Target Broker || After current weight ||
+ * ||----------1-----------||------ {5, 1, 1} ------||----Broker1----||----- {-2, 1, 1} -----||
+ * ||----------2-----------||------ {3, 2, 2} ------||----Broker1----||----- {-4, 2, 2} -----||
+ * ||----------3-----------||------ {1, 3, 3} ------||----Broker2----||----- { 1,-4, 3} -----||
+ * ||----------4-----------||------ {6,-3, 4} ------||----Broker1----||----- {-1,-3, 4} -----||
+ * ||----------5-----------||------ {4,-2, 5} ------||----Broker3----||----- { 4,-2,-2} -----||
+ * ||----------6-----------||------ {9,-1,-1} ------||----Broker1----||----- { 2,-1,-1} -----||
+ * ||----------7-----------||------ {7, 0, 0} ------||----Broker1----||----- { 0, 0, 0} -----||
+ * ||======================||=======================||===============||======================||
+ */
+public final class SmoothWeightRoundRobin
```

Comment:
我大概看了一下邏輯，並且把全部用read-only + lambda的方式改寫，這樣的好處是
1. 之後要做immutable的話很好用，只要加上final就可以檢查哪一個物件被更改了
2. 做同步也會比較簡單，因為每個物件的修改都只有"="時會被修改，不會有那種“改變物件內部資料”的行為
3. 閱讀起來也會更直覺

```java
package org.astraea.partitioner.smooth;

import java.util.Comparator;
import java.util.Map;
import java.util.stream.Collectors;
import org.astraea.cost.Periodic;
import org.astraea.cost.broker.CostUtils;

/**
 * Given initial key-score pair, it will output a preferred key with the highest current weight. The
 * current weight of the chosen key will decrease the sum of effective weight. And all current
 * weight will increment by its effective weight. It may result in "higher score with higher chosen
 * rate". For example:
 *
 * <p>||========================================||============================================||
 * ||------------ Broker in cluster ------------||------------- Effective weight -------------||
 * ||------------------Broker1------------------||-------------------- 5 ---------------------||
 * ||------------------Broker2------------------||-------------------- 1 ---------------------||
 * ||------------------Broker3------------------||-------------------- 1 ---------------------||
 * ||===========================================||============================================||
 *
 * <p>||===================||=======================||===============||======================||
 * ||--- Request Number ---|| Before current weight || Target Broker || After current weight ||
 * ||----------1-----------||------ {5, 1, 1} ------||----Broker1----||----- {-2, 1, 1} -----||
 * ||----------2-----------||------ {3, 2, 2} ------||----Broker1----||----- {-4, 2, 2} -----||
 * ||----------3-----------||------ {1, 3, 3} ------||----Broker2----||----- { 1,-4, 3} -----||
 * ||----------4-----------||------ {6,-3, 4} ------||----Broker1----||----- {-1,-3, 4} -----||
 * ||----------5-----------||------ {4,-2, 5} ------||----Broker3----||----- { 4,-2,-2} -----||
 * ||----------6-----------||------ {9,-1,-1} ------||----Broker1----||----- { 2,-1,-1} -----||
 * ||----------7-----------||------ {7, 0, 0} ------||----Broker1----||----- { 0, 0, 0} -----||
 * ||======================||=======================||===============||======================||
 */
public final class SmoothWeightRoundRobin
    extends Periodic<SmoothWeightRoundRobin.EffectiveWeightResult> {
  private EffectiveWeightResult effectiveWeightResult;
  public Map<Integer, Double> currentWeight;

  public SmoothWeightRoundRobin(Map<Integer, Double> effectiveWeight) {
    effectiveWeightResult =
        new EffectiveWeightResult(
            effectiveWeight.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, ignored -> 1.0)));
    currentWeight =
        effectiveWeight.entrySet().stream()
            .collect(Collectors.toMap(Map.Entry::getKey, ignored -> 0.0));
  }

  public synchronized void init(Map<Integer, Double> brokerScore) {
    effectiveWeightResult =
        tryUpdate(
            () -> {
              var normalizationLoad = CostUtils.normalize(brokerScore);
              return new EffectiveWeightResult(
                  this.effectiveWeightResult.effectiveWeight.entrySet().stream()
                      .collect(
                          Collectors.toMap(
                              Map.Entry::getKey,
                              entry -> {
                                var nLoad = normalizationLoad.get(entry.getKey());
                                var weight =
                                    entry.getValue()
                                        * (nLoad.isNaN()
                                            ? 1.0
                                            : ((nLoad + 1) > 0 ? nLoad + 1 : 0.1));
                                if (weight > 2.0) return 2.0;
                                return Math.max(weight, 0.0);
                              })));
            },
            10);
  }

  /**
   * Get the preferred ID, and update the state.
   *
   * @return the preferred ID
   */
  public synchronized int getAndChoose() {
    this.currentWeight =
        this.currentWeight.entrySet().stream()
            .collect(
                Collectors.toMap(
                    Map.Entry::getKey,
                    e -> e.getValue() + effectiveWeightResult.effectiveWeight.get(e.getKey())));
    var maxID =
        this.currentWeight.entrySet().stream()
            .max(Comparator.comparingDouble(Map.Entry::getValue))
            .map(Map.Entry::getKey)
            .orElse(0);
    this.currentWeight =
        this.currentWeight.entrySet().stream()
            .collect(
                Collectors.toMap(
                    Map.Entry::getKey,
                    e ->
                        e.getKey().equals(maxID)
                            ? e.getValue() - effectiveWeightResult.effectiveWeightSum
                            : e.getValue()));
    return maxID;
  }

  public static class EffectiveWeightResult {
    private final Map<Integer, Double> effectiveWeight;
    private final double effectiveWeightSum;

    EffectiveWeightResult(Map<Integer, Double> effectiveWeight) {
      this.effectiveWeight = effectiveWeight;
      this.effectiveWeightSum = effectiveWeight.values().stream().mapToDouble(i -> i).sum();
    }
  }
}

```

Reply:
lambda我看出來是把我幾個分散的變量全部都揉到一個lambda裏去了。
但read-only的方式我不太明白，是指都只有"="時會被修改嗎。
說實話我認知裏的read-only基本指final。